packages/@aws-cdk/aws-ec2/lib/cfn-init.ts

```diff
@@ -0,0 +1,362 @@
+  /**
+   * Build a new config from a set of Init Elements
+   */
+  public static fromElements(...elements: InitElement[]): CloudFormationInit {
+    return CloudFormationInit.fromConfig(new InitConfig(elements));
+  }
+
+  /**
+   * Use an existing InitConfig object as the default and only config
+   */
+  public static fromConfig(config: InitConfig): CloudFormationInit {
+    return CloudFormationInit.fromConfigSets({
+      configSets: {
+        default: ['config'],
+      },
+      configs: { config },
+    });
+  }
+
+  /**
+   * Build a CloudFormationInit from config sets
+   */
+  public static fromConfigSets(props: ConfigSetProps): CloudFormationInit {
+    return new CloudFormationInit(props.configSets, props.configs);
+  }
+
+  private readonly _configSets: Record<string, string[]> = {};
+  private readonly _configs: Record<string, InitConfig> = {};
+
+  protected constructor(configSets: Record<string, string[]>, configs: Record<string, InitConfig>) {
+    Object.assign(this._configSets, configSets);
+    Object.assign(this._configs, configs);
+  }
+
+  /**
+   * Add a config with the given name to this CloudFormationInit object
+   */
+  public addConfig(configName: string, config: InitConfig) {
+    if (this._configs[configName]) {
+      throw new Error(`CloudFormationInit already contains a config named '${configName}'`);
+    }
+    this._configs[configName] = config;
+  }
+
+  /**
+   * Add a config set with the given name to this CloudFormationInit object
+   *
+   * The new configset will reference the given configs in the given order.
+   */
+  public addConfigSet(configSetName: string, configNames: string[] = []) {
+    if (this._configSets[configSetName]) {
+      throw new Error(`CloudFormationInit already contains a configSet named '${configSetName}'`);
+    }
+
+    const unk = configNames.filter(c => !this._configs[c]);
+    if (unk.length > 0) {
+      throw new Error(`Unknown configs referenced in definition of '${configSetName}': ${unk}`);
+    }
+
+    this._configSets[configSetName] = [...configNames];
+  }
+
+  /**
+   * Attach the CloudFormation Init config to the given resource
+   *
+   * This method does the following:
+   *
+   * - Renders the `AWS::CloudFormation::Init` object to the given resource's
+   *   metadata, potentially adding a `AWS::CloudFormation::Authentication` object
+   *   next to it if required.
+   * - Updates the instance role policy to be able to call the APIs required for
+   *   `cfn-init` and `cfn-signal` to work, and potentially add permissions to download
+   *   referenced asset and bucket resources.
+   * - Updates the given UserData with commands to execute the `cfn-init` script.
+   *
+   * As an app builder, you shouldn't need to use this API directly. Instead,
+   * use `instance.applyCloudFormationInit()` or
+   * `autoScalingGroup.applyCloudFormationInit()`.
+   */
+  public attach(attachedResource: CfnResource, attachOptions: AttachInitOptions) {
```

Comment:
It doesn't look like we expect customers to call this API? If so, hide this and `AttachInitOptions` from the public API.